stripe/src/main/java/com/stripe/android/IssuingCardPinService.java

```diff
@@ -0,0 +1,309 @@
+                            "The one-time code has expired",
+                            null);
+                }
+                if ("incorrect_code".equals(e.getErrorCode())) {
+                    listener.onError(
+                            CardPinActionError.ONE_TIME_CODE_INCORRECT,
+                            "The one-time code was incorrect",
+                            null);
+                }
+                if ("too_many_attempts".equals(e.getErrorCode())) {
+                    listener.onError(
+                            CardPinActionError.ONE_TIME_CODE_TOO_MANY_ATTEMPTS,
+                            "The verification challenge was attempted too many times",
+                            null);
+                } else {
+                    listener.onError(
+                            CardPinActionError.UNKNOWN_ERROR,
+                            "An error occurred retrieving the PIN",
+                            e);
+                }
+            } catch (APIConnectionException |
+                    APIException |
+                    AuthenticationException |
+                    CardException e) {
+                listener.onError(
+                        CardPinActionError.UNKNOWN_ERROR,
+                        "An error occurred retrieving the PIN",
+                        e);
+            }
+        }
+    }
+
+    @Override
+    public void onKeyError(@Nullable String operationId,
+                           int errorCode,
+                           @Nullable String errorMessage) {
+
+        IssuingCardPinUpdateListener updateListener = mUpdateListeners.remove(operationId);
+        IssuingCardPinRetrievalListener retrievalListener = mRetrievalListeners.remove(operationId);
+        if (retrievalListener != null) {
+            retrievalListener.onError(
+                    CardPinActionError.EPHEMERAL_KEY_ERROR,
+                    errorMessage,
+                    null);
+        } else if (updateListener != null) {
+            updateListener.onError(
+                    CardPinActionError.EPHEMERAL_KEY_ERROR,
+                    errorMessage,
+                    null);
+        }
+    }
+
+    public enum CardPinActionError {
+        UNKNOWN_ERROR,
+        EPHEMERAL_KEY_ERROR,
+        ONE_TIME_CODE_INCORRECT,
+        ONE_TIME_CODE_EXPIRED,
+        ONE_TIME_CODE_TOO_MANY_ATTEMPTS,
+    }
+
+    public interface IssuingCardPinRetrievalListener {
+        void onIssuingCardPinRetrieved(@NonNull String pin);
+
+        void onError(
+                @NonNull CardPinActionError errorCode,
+                @Nullable String errorMessage,
+                @Nullable Throwable exception);
+    }
+
+    public interface IssuingCardPinUpdateListener {
+        void onIssuingCardPinUpdated();
+
+        void onError(
+                @NonNull CardPinActionError errorCode,
+                @Nullable String errorMessage,
+                @Nullable Throwable exception);
```

Comment:
Is there value in passing the errorMessage and exception? Is errorCode sufficient?

Reply:
hm I see what you mean. Here the codes enum names are possibly explicit enough, but in general it's nice to have a little bit more room to describe the problem. The `UNKNOWN_ERROR` could use some better messages at one point.
I can remove the string if you feel strongly about it, that'd be 100% OK.